File: runtime-testsuite/test/org/antlr/v5/test/runtime/java/helpers/TreeShapeListener.java
package org.antlr.v5.test.runtime.java.helpers;

import org.antlr.v5.runtime.core.context.ParserRuleContext;
import org.antlr.v5.runtime.core.tree.*;

public class TreeShapeListener implements ParseTreeListener {
	public static final TreeShapeListener INSTANCE = new TreeShapeListener();

	@Override public void visitTerminal(TerminalNode node) { }
	@Override public void visitErrorNode(ErrorNode node) { }
	@Override public void exitEveryRule(ParserRuleContext ctx) { }

	@Override
	public void enterEveryRule(ParserRuleContext ctx) {
		for (int i = 0; i < ctx.getChildCount(); i++) {
			ParseTree parent = ctx.getChild(i).getParent();
			if (!(parent instanceof RuleNode) || ((RuleNode)parent).getRuleContext() != ctx) {
				throw new IllegalStateException("Invalid parse tree shape detected.");
			}
		}
	}
}

File: runtime-testsuite/test/org/antlr/v5/test/runtime/kotlin/helpers/TreeShapeListener.kt
package org.antlr.v5.test.runtime.kotlin.helpers

import org.antlr.v5.runtime.core.context.ParserRuleContext
import org.antlr.v5.runtime.core.tree.ErrorNode
import org.antlr.v5.runtime.core.tree.ParseTreeListener
import org.antlr.v5.runtime.core.tree.RuleNode
import org.antlr.v5.runtime.core.tree.TerminalNode


class TreeShapeListener : ParseTreeListener {
    override fun visitTerminal(node: TerminalNode) {}
    override fun visitErrorNode(node: ErrorNode) {}
    override fun exitEveryRule(ctx: ParserRuleContext) {}

    override fun enterEveryRule(ctx: ParserRuleContext) {
        for (i in 0 until ctx.childCount) {
            val parent = ctx.getChild(i)!!.getParent()
            check(!(parent !is RuleNode || parent.ruleContext !== ctx)) { "Invalid parse tree shape detected." }
        }
    }

    companion object {
        val INSTANCE: TreeShapeListener = TreeShapeListener()
    }
}
